### DISCUSSION TITLE
does $nextTick fire before the DOM repaints?

### DISCUSSION THREAD

**Discussion by nasheomirro** [alpinejs/alpine#4092] [1. Help] [answered]
I'm using x-for and trying to do FLIP animations where the items that are re-organized will animate to their new locations:
```html
<template x-for="item in items" :key="item.id">
   <div x-init="item.ref = $el" x-text="item.text"></div>
</template>
```
```js
const initialStates = getBoundingClientRects(this.items); // get initial rects for calculating inverse/transform
this.items.filter(item => item.category === 'cats'); // filter the items to remove some elements

await this.$nextTick(); // wait for alpine to update the DOM because of x-for

const finalStates = getBoundingClientRects(this.items); // get the final rects for calculation
const transforms = calculateInverse(items, initialStates, finalStates); // set and calculate the transforms

// then use requestAnimationFrame to strip the transforms out, leading to the elements transitioning to their new locations
requestAnimationFrame(() => playTransformations(transforms));
```
This works better than I expected, but the one problem is there is a slight frame where we could see the elements are already in their final transtitions before doing the animation, I think it's because `nextTick()` doesn't stop the DOM from repainting, so the DOM had a chance to paint before applying the transforms.

Is this what's happening here or is my hunch wrong? And if `nextTick()` does let the browser repaint is there a way I could let alpine update the DOM but before it repaints the screen similar to `useLayoutEffect()` in react?

**Comment by ekwoka** [comment, marked as answer]
Next tick fires at the start of the next frame, after paint

**Comment by nasheomirro** [comment]
I was able to get this working by using `queueMicroTask()`. I have no knowledge of how Alpine does renders, but I'm assuming they use similar or even the same function for rendering. All I know is that it seems the DOM has already been updated by Alpine before our callback to `queueMicroTask()` gets executed.

Played around with this and found out that the order of where you place `queueMicroTask()` matters, if you place it before telling Alpine to, for example, filter items from state, then you might end up executing your callback in the middle of Alpine's changes, again though I'm not sure if this would work consistently.

**Reply by ekwoka** [reply]
You probably just want to use `requestAnimationFrame`. So that you will run after the microtaskqueue but before the screen paints.

alternatively, use css animations.
